java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -206,4 +220,41 @@ private static boolean isConstructorWithParameters(MethodTree methodTree) {
   private static boolean isNotMethodFromSerializable(MethodTree methodTree, Symbol symbol) {
     return methodTree.is(Tree.Kind.METHOD) && !SerializableContract.SERIALIZABLE_CONTRACT_METHODS.contains(symbol.name());
   }
+
+  private static class CheckAnnotationsForMethodNames extends BaseTreeVisitor {
+
+    public CheckAnnotationsForMethodNames(Set<String> methodNames) {
+      this.methodNames = methodNames;
+    }
+
+    private Set<String> methodNames;
+
+    private static boolean isNameIndicatingMethod(String name) {
+      return name.toLowerCase(Locale.getDefault()).contains("method");
+    }
+
+    private void removeMethodName(LiteralTree literal) {
+      methodNames.remove(removeQuotes(literal.value()));
+    }
+
+    private static String removeQuotes(String withQuotes) {
+      return withQuotes.substring(1, withQuotes.length() - 1);
+    }
+
+    @Override
+    public void visitAnnotation(AnnotationTree annotationTree) {
+      var isMethodAnnotation = isNameIndicatingMethod(annotationTree.annotationType().symbolType().name());
+      for (var arg : annotationTree.arguments()) {
+        if (arg.is(Tree.Kind.STRING_LITERAL)) {
+          if (isMethodAnnotation) {
+            removeMethodName((LiteralTree) arg);
+          }
+        } else if (arg instanceof AssignmentExpressionTree asgn && asgn.expression().is(Tree.Kind.STRING_LITERAL) && (
+          isMethodAnnotation || isNameIndicatingMethod(((IdentifierTree) asgn.variable()).name())
+        )) {
+          removeMethodName((LiteralTree) asgn.expression());
```

Comment:
Not a super strong opinion here, but maybe it could be more comprehensible like
```suggestion
 else if (isAnnotationArgumentIndicatingMethod(arg, isMethodAnnotation)) {
          removeMethodName((LiteralTree) ((AssignmentExpressionTree)arg).expression());
        }
```

with the logic separated in another boolean method
```
    private static boolean isAnnotationArgumentIndicatingMethod(ExpressionTree arg, boolean isMethodAnnotation) {
      return arg instanceof AssignmentExpressionTree asgn && asgn.expression().is(Tree.Kind.STRING_LITERAL) && (
        isMethodAnnotation || isNameIndicatingMethod(((IdentifierTree) asgn.variable()).name())
      );
    }
```